File: Trabalho 1/Radixsort.c
/*
 * Guilherme Araújo Mendes de Souza - 156437
 * UNIFESP - ICT
 * AED 2
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int getMax(int vetor[], int n) {
  int maximo = vetor[0];
  for (int i = 1; i < n; i++)
    if (vetor[i] > maximo)
      maximo = vetor[i];
  return maximo;
}

void countingSort(int vetor[], int tamanho, int lugar) {
  int saida[tamanho];
  int maximo = (vetor[0] / lugar) % 10;

  for (int i = 1; i < tamanho; i++) {
    if (((vetor[i] / lugar) % 10) > maximo)
      maximo = (vetor[i] / lugar) % 10;
  }
  int contador[maximo + 1];

  for (int i = 0; i <= maximo; ++i)
    contador[i] = 0;

  for (int i = 0; i < tamanho; i++)
    contador[(vetor[i] / lugar) % 10]++;
    
  for (int i = 1; i <= maximo; i++)
    contador[i] += contador[i - 1];

  for (int i = tamanho - 1; i >= 0; i--) {
    saida[contador[(vetor[i] / lugar) % 10] - 1] = vetor[i];
    contador[(vetor[i] / lugar) % 10]--;
  }

  for (int i = 0; i < tamanho; i++)
    vetor[i] = saida[i];
}

void radixsort(int tamanho, int vetor[]) {
  int maximo = getMax(vetor, tamanho);

  for (int lugar = 1; maximo / lugar > 0; lugar *= 10)
    countingSort(vetor, tamanho, lugar);
}

int main() {
    clock_t inicio, fim;
    FILE *arquivo;
    char nome[25];
    int tamanhoVetor;
    int i = 0;

    printf("Informe o nome do arquivo: ");
    scanf("%s", nome);

    printf("\nInforme o tamanho do arquivo: ");
    scanf("%d", &tamanhoVetor);

    int *vetor = (int *)malloc(tamanhoVetor * sizeof(int));

    if (vetor == NULL) {
        printf("\nFalha na alocação de memória.\n");
        return 1;
    }

    arquivo = fopen(nome, "r");

    if (arquivo == NULL) {
        printf("Erro ao tentar abrir o arquivo\n");
        return 1;
    }

    while (i < tamanhoVetor && fscanf(arquivo, "%d", &vetor[i]) != EOF) {
        i++;
    }
    
    fclose(arquivo);

    inicio = clock();
    radixsort(tamanhoVetor, vetor);
    fim = clock();

    double tempo_decorrido = (((double) (fim - inicio)) / CLOCKS_PER_SEC) * 1000;
    printf("Tempo de CPU com %d posições no vetor: %f milissegundos\n", tamanhoVetor, tempo_decorrido);

    free(vetor);
    return 0;
}

File: Trabalho 1/Radix_Sort.py
'''
Guilherme Araújo Mendes de Souza - 156437
UNIFESP - ICT
AED 2
'''
import time

def countingSort(vetor, lugar):
    tamanho = len(vetor)
    saida = [0] * tamanho
    contador = [0] * 10

    for i in range(0, tamanho):
        indice = (vetor[i] // lugar) % 10
        contador[indice] += 1

    for i in range(1, 10):
        contador[i] += contador[i - 1]

    i = tamanho - 1
    while i >= 0:
        indice = (vetor[i] // lugar) % 10
        saida[contador[indice] - 1] = vetor[i]
        contador[indice] -= 1
        i -= 1

    for i in range(0, tamanho):
        vetor[i] = saida[i]

def radixSort(vetor):
    maximo_elemento = max(vetor)
    lugar = 1

    while maximo_elemento // lugar > 0:
        countingSort(vetor, lugar)
        lugar *= 10

nome = input('Informe o nome do arquivo: ')
tamanho_vetor = int(input('Informe o tamanho do arquivo: '))
vetor = []

arquivo = open(nome, 'r')

for linha in arquivo:
    vetor.append(int(linha.strip()))  # Converte a linha para inteiro

arquivo.close()

inicio = time.process_time()
radixSort(vetor)
fim = time.process_time()

tempo_decorrido = (fim - inicio) * 1000  # Multiplica por 1000 para obter milissegundos
print(f'Tempo de CPU com {tamanho_vetor} posições no vetor: {tempo_decorrido} milissegundos')